storage/gcloud/aio/storage/storage.py

```diff
@@ -485,18 +490,58 @@ def _get_stream_len(stream: IO[AnyStr]) -> int:
             stream.seek(current)
 
     @staticmethod
-    def _preprocess_data(data: Any) -> IO[Any]:
+    def _compress_file_in_chunks(input_stream: Union[IO[AnyStr], io.IOBase],
+                                 output_stream: io.IOBase,
+                                 chunk_size: int = 8192) -> None:
+        """
+        Reads the contents of input_stream and writes it gzip-compressed to
+        output_stream in chunks. The chunk size is 8Kb by default, which is a
+        standard filesystem block size.
+        """
+        with gzip.open(output_stream, 'wb') as f_out:
+            while True:
+                chunk = input_stream.read(chunk_size)
+                if not chunk:
+                    break
+                if isinstance(chunk, str):
+                    chunk_bytes = chunk.encode('utf-8')
+                else:
+                    chunk_bytes = chunk
+                # At this point we assume that chunk is a bytes object
```

Comment:
Rather than assuming, we can verify this by pre-declaring our type hint.

```suggestion
                chunk_bytes: bytes
                if isinstance(chunk, str):
                    chunk_bytes = chunk.encode('utf-8')
                else:
                    chunk_bytes = chunk
```

Reply:
Even with the type hint, it would still be an assumption. This is because the `io.IOBase` interface doesn't have a `read` method with a type hint that specified that the return type is either `str` or `bytes`, even though all the implementations I'm aware of return one of those 2. I find this part of Python's typing system and its inconsistency with the `typing.IO` classes very confusing.

From the docs:

> Even though [IOBase](https://docs.python.org/3/library/io.html#io.IOBase) does not declare read() or write() because their signatures will vary, implementations and clients should consider those methods part of the interface. Also, implementations may raise a [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError) (or [UnsupportedOperation](https://docs.python.org/3/library/io.html#io.UnsupportedOperation)) when operations they do not support are called.

Reply:
By moving `_compress_file_in_chunks` out of `_preprocess_data` I was able to change the typing hints so `input_stream` is now a `IO[AnyStr]`, which does bound the return type of `read`